pkg/storage/unified/apistore/stream.go

```diff
@@ -35,25 +48,30 @@ func (d *streamDecoder) toObject(w *resource.WatchEvent_Resource) (runtime.Objec
 	return obj, err
 }
 
+// nolint: gocyclo // we may be able to simplify this in the future, but this is a complex function by nature
 func (d *streamDecoder) Decode() (action watch.EventType, object runtime.Object, err error) {
+	d.done.Add(1)
```

Comment:
Drive-by comment: I can't see where you start the goroutine, but take note of this in [the WaitGroup docs](https://pkg.go.dev/sync#WaitGroup.Add): 
> Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for.